p4-fusion/commands/change_list.cc

```diff
@@ -108,49 +104,88 @@ void ChangeList::StartDownload(P4API* p4, const BranchSet& branchSet, const int&
 
 	// Flush any remaining files that were smaller in number than the total batch size.
 	// Additionally, signal the batch processing end.
-	Flush(p4, printBatchFiles, printBatchFileData);
+	Flush(p4, printBatchFileData);
 	*downloadJobsCompleted = true;
 	commitCV->notify_all();
 }
 
-void ChangeList::Flush(P4API* p4, std::shared_ptr<std::vector<std::string>> printBatchFiles, std::shared_ptr<std::vector<FileData*>> printBatchFileData)
+void ChangeList::Flush(P4API& p4, const std::shared_ptr<std::vector<FileData*>>& printBatchFileData)
 {
-	// Only perform the batch processing when there are files to process.
-	if (!printBatchFileData->empty())
-	{
-		const PrintResult& printData = p4->PrintFiles(*printBatchFiles);
-		if (printData.HasError())
-		{
-			throw printData.PrintError();
-		}
+	MTR_SCOPE("ChangeList", __func__);
 
-		for (int i = 0; i < printBatchFiles->size(); i++)
-		{
-			printBatchFileData->at(i)->MoveContentsOnceFrom(printData.GetPrintData().at(i).contents);
-		}
+	std::vector<std::string> fileRevisions;
+	for (auto fileData : *printBatchFileData)
+	{
+		fileRevisions.push_back(fileData->GetDepotFile() + "#" + fileData->GetRevision());
+	}
 
-		(*filesDownloaded) += printBatchFiles->size();
+	// Only perform the batch processing when there are files to process.
+	if (fileRevisions.empty())
+	{
+		return;
```

Comment:
`fileRevisions.size()` == `printBatchFileData.size()`, so this check should be moved earlier before the copy.